### ISSUE TITLE
[Question] sliding window attention + dynamic input shapes

### ISSUE THREAD

**Issue by pfeatherstone** [pytorch-labs/attention-gym#138]
I want to use a sliding window attention with any input length. Currently my attention layer is as follows:
```Py
from   functools import partial
import torch
import torch.nn as nn
import torch.nn.functional as F
from   torch.nn.attention.flex_attention import flex_attention, create_block_mask
from   einops import rearrange

def sliding_window(window_size, b, h, q_idx, kv_idx):
    hlen  = window_size // 2
    mask0 = q_idx - kv_idx <= hlen
    mask1 = kv_idx - q_idx <= hlen
    return mask0 & mask1
    

class Attention(nn.Module):
    def __init__(self, dim, num_heads, window_size):
        super().__init__()
        dim_head    = dim // num_heads
        self.H      = num_heads
        self.to_qkv = nn.Linear(dim, 3*num_heads*dim_head, bias=False)
        self.to_out = nn.Linear(num_heads*dim_head, dim,   bias=False)
        self.mask   = partial(sliding_window, window_size)
    
    def forward(self, x):
        q, k, v = map(lambda t: rearrange(t, 'b t (h d) -> b h t d', h=self.H), self.to_qkv(x).chunk(3, -1))
        mask    = create_block_mask(self.mask, B=None, H=None, Q_LEN=x.shape[1], KV_LEN=x.shape[1])
        x       = flex_attention(q, k, v, block_mask=mask)
        x       = self.to_out(rearrange(x, 'b h t d -> b t (h d)'))
        return x
```

In order for it to work with any sequence length, i have to create the block mask in the `forward()` function. 

- Is this the best way to do this? 
- Will it recompile automatically every time  `x.shape[1]` changes?
- Will this work with ONNX export ?

**Comment by joydddd** [comment]
I assume your window size is a constant. 

You can create a `block_mask` that works for the MAX sequence length you'd expect and reuse that mask. When Q K V shape is smaller than the input mask, it would use the top left corner of it.

**Comment by pfeatherstone** [comment]
Ok cool. What about onnx export ?

**Comment by drisspg** [comment]
I have not tested it but I am 99.999% sure this will not work w/ ONNX
